controllers/apim/ratelimitpolicy_wasm_plugins.go

```diff
@@ -116,45 +120,52 @@ func (r *RateLimitPolicyReconciler) gatewayWASMPlugin(ctx context.Context, gw *g
 }
 
 // returns nil when there is no rate limit policy to apply
-func (r *RateLimitPolicyReconciler) wasmPluginConfig(ctx context.Context, gw *gatewayapiv1alpha2.Gateway, rlpRefs []client.ObjectKey) (*rlptools.WASMPlugin, error) {
+func (r *RateLimitPolicyReconciler) wasmPluginConfig(ctx context.Context,
+	gw rlptools.GatewayWrapper, rlpRefs []client.ObjectKey) (*rlptools.WASMPlugin, error) {
 	logger, _ := logr.FromContext(ctx)
-	gwKey := client.ObjectKeyFromObject(gw)
-	// Load all relevant rate limit policies
+
 	routeRLPList := make([]*apimv1alpha1.RateLimitPolicy, 0)
-	gwRLPList := make([]*apimv1alpha1.RateLimitPolicy, 0)
+	var gwRLP *apimv1alpha1.RateLimitPolicy
 	for _, rlpKey := range rlpRefs {
 		rlp := &apimv1alpha1.RateLimitPolicy{}
 		err := r.Client().Get(ctx, rlpKey, rlp)
-		logger.V(1).Info("gatewayWASMPlugin", "get rlp", rlpKey, "err", err)
+		logger.V(1).Info("wasmPluginConfig", "get rlp", rlpKey, "err", err)
 		if err != nil {
 			return nil, err
 		}
 
 		if rlp.IsForHTTPRoute() {
 			routeRLPList = append(routeRLPList, rlp)
 		} else if rlp.IsForGateway() {
-			gwRLPList = append(gwRLPList, rlp)
+			if gwRLP == nil {
+				gwRLP = rlp
+			} else {
+				return nil, fmt.Errorf("wasmPluginConfig: multiple gateway RLP found and only one expected. rlp keys: %v", rlpRefs)
+			}
```

Comment:
Maybe in a future refactor we could extract this sorting of the RLPs in a reusable fn, `sortRLPs` or something similar, to use also in `gatewayLimits`